modules/caddyhttp/server.go

```diff
@@ -579,21 +614,25 @@ func (slc ServerLogConfig) getLoggerName(host string) string {
 
 // PrepareRequest fills the request r for use in a Caddy HTTP handler chain. w and s can
 // be nil, but the handlers will lose response placeholders and access to the server.
-func PrepareRequest(r *http.Request, repl *caddy.Replacer, w http.ResponseWriter, s *Server) *http.Request {
-	// set up the context for the request
-	ctx := context.WithValue(r.Context(), caddy.ReplacerCtxKey, repl)
-	ctx = context.WithValue(ctx, ServerCtxKey, s)
-	ctx = context.WithValue(ctx, VarsCtxKey, make(map[string]interface{}))
-	ctx = context.WithValue(ctx, routeGroupCtxKey, make(map[string]struct{}))
-	var url2 url.URL // avoid letting this escape to the heap
-	ctx = context.WithValue(ctx, OriginalRequestCtxKey, originalRequest(r, &url2))
-	r = r.WithContext(ctx)
+// Along with the new request it returns the RequestContext stored inside so that
+// the caller may set additional context values if needed, which should be preferred
+// over context.WithValue and r.WithContext which would be slower and result in more allocations.
+func PrepareRequest(r *http.Request, repl *caddy.Replacer, w http.ResponseWriter, s *Server) (
```

Comment:
I think `PrepareRequest` is only exported to be used in tests, right? Maybe we should say as much in the godoc comment, to make sure nobody tries to use this at runtime other than `caddyhttp`.

Reply:
If I remember correctly, PrepareRequest is indeed currently only used in tests though I can imagine a usecase where somebody wants a custom Caddy app which receives requests in some non-http form and then serves them using Caddy's http handlers. In the way PrepareRequest is currently implemented, it does not enforce the caller to return the RequestContext to the pool or even use it, so we might as well leave it exported without "marking" it as "for internal use only", no?